### PR TITLE
Data access and listing all books

### PR DESCRIPTION
**Pull request by kwunkiuma**

### CODE REVIEW

File: Bookish.DataAccess/DataHelper.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using Dapper;
+
+namespace Bookish.DataAccess
+{
+    public class DataHelper
```
Comment (FridaTveit) [inline Bookish.DataAccess/DataHelper.cs]: Classes like this are usually called services, e.g. `BookishService` 🙂

File: Bookish.DataAccess/DataHelper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        private static readonly IDbConnection DbConnection = new SqlConnection(@"Server = localhost; Database = Bookish; Trusted_Connection = True;");
```
Comment (FridaTveit) [inline Bookish.DataAccess/DataHelper.cs]: I wouldn't recommend having this as a static variable. I would make this an instance variable and make the class have only instance methods 🙂

File: Bookish.DataAccess/DataHelper.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+        public static IEnumerable<T> ExecuteQuery<T>(string query)
```
Comment (FridaTveit) [inline Bookish.DataAccess/DataHelper.cs]: This class should be fully responsible for interacting with the database, therefore it should handle creating the query 🙂

**Review by FridaTveit — COMMENTED** [review]
A few comments otherwise looks good 🙂 Just to check, have you put your sql scripts up for review? If not, please could you add them to this PR so that I can review them and the db structure in general 🙂

File: Bookish.ConsoleApp/Program.cs [hunk 4]
```diff
@@ -5,15 +5,18 @@ namespace Bookish.ConsoleApp
...
         {
-            var response = DataHelper.ExecuteQuery<Book>(@"SELECT * FROM BOOKS");
+            var response = service.GetBooks();
             Console.WriteLine(string.Join("\n", response));
         }
 
         static void Main()
         {
-            ListBooks();
+            var service = new BookishService();
```
Comment (FridaTveit) [inline Bookish.ConsoleApp/Program.cs]: You could just create this variable in the ListBooks method? No need to pass it in as you're not using it again 🙂

**Comment by kwunkiuma** [comment]
Storing the password as plaintext is only for this exercise!

File: Bookish.DataAccess/CreateAccountTable.sql [hunk 5]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE dbo.Accounts (
+	Username varchar(20) PRIMARY KEY NOT NULL,
+	Pass varchar(20) NOT NULL
```
Comment (FridaTveit) [inline Bookish.DataAccess/CreateAccountTable.sql]: I guess you could use the username as the id. But it's more common to have a separate user id 🙂 

Once you get to adding the website step the framework will handle the login for you. I wouldn't bother storing a password for now 🙂

File: Bookish.DataAccess/CreateAccountTable.sql [hunk 6]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE dbo.Accounts (
```
Comment (FridaTveit) [inline Bookish.DataAccess/CreateAccountTable.sql]: Minor but you shouldn't need `dbo` 🙂

File: Bookish.DataAccess/CreateBooksTable.sql [hunk 7]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE dbo.Books (
+	ISBN char(13) PRIMARY KEY NOT NULL,
+	Title varchar(50) NOT NULL,
+	Author varchar(50) NOT NULL,
+	TotalCopies int NOT NULL,
+	AvailableCopies int NOT NULL
```
Comment (FridaTveit) [inline Bookish.DataAccess/CreateBooksTable.sql]: Hm, this doesn't help you keep track of which copies are available and which ones are on loan with which users. Having a separate BookCopies table or similar can probably help with that 🙂

File: Bookish.DataAccess/BookishService.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using Dapper;
+
+namespace Bookish.DataAccess
+{
+    public class BookishService
+    {
+        private readonly IDbConnection dbConnection = new SqlConnection(@"Server = localhost; Database = Bookish; Trusted_Connection = True;");
```
Comment (FridaTveit) [inline Bookish.DataAccess/BookishService.cs]: You don't need the `@` I believe 🙂

**Review by FridaTveit — COMMENTED** [review]
A few suggestions, otherwise looks good. It would probably also be useful with a script to insert some data into your database? 🙂

**Comment by kwunkiuma** [comment]
What would be better out of these approaches?
1. The current approach where there is a Loans table that maps copies to lenders
2. Store lender and due date on every copy; storing null when the copy is available.

File: Bookish.DataAccess/Account.cs [hunk 9]
```diff
@@ -1,13 +1,13 @@
 ﻿namespace Bookish.DataAccess
 {
-    public class Account
+    class Account
```
Comment (FridaTveit) [inline Bookish.DataAccess/Account.cs]: Do you actually need any of these classes for now other than book? 🙂

File: Bookish.DataAccess/CreateLoansTable.sql [hunk 10]
```diff
@@ -1,7 +1,7 @@
-CREATE TABLE dbo.Loans (
-	LoanID int PRIMARY KEY NOT NULL,
-	ISBN char(13) FOREIGN KEY REFERENCES Books(ISBN) NOT NULL,
-	Username varchar(20) FOREIGN KEY REFERENCES Accounts(Username) NOT NULL,
+CREATE TABLE Loans (
+	CopyID int FOREIGN KEY REFERENCES BookCopies(CopyID) NOT NULL,
+	LenderID int FOREIGN KEY REFERENCES Accounts(UserID) NOT NULL,
 	DueDate date NOT NULL,
+	CONSTRAINT PK_Loan PRIMARY KEY (CopyID, LenderID)
```
Comment (FridaTveit) [inline Bookish.DataAccess/CreateLoansTable.sql]: The copyId should be enough for a primary key, as it can't be loaned to more than one person at once 🙂

**Review by FridaTveit — COMMENTED** [review]
Looks good! I've left a few comments. 

Good question about storing lender and due date on the copy. I think the current approach is preferrable, because with the other approach your db schema would allow a copy to have a lender but no due date or vice versa, which it shouldn't allow 🙂

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂

**Pull request merged**